src/dataflow/src/render/reduce.rs

```diff
@@ -502,12 +508,66 @@ impl ReducePlan {
                     ));
                 }
                 // Now we need to collate them together.
-                build_collation(to_collate, expr.aggregate_types, &mut collection.scope())
+                build_collation(to_collate, expr.aggregate_types, &mut collection.scope()).into()
             }
+        };
+        arrangement_or_bundle.to_bundle(key_arity, err_input)
+    }
+}
+
+enum ArrangementOrBundle<G, T>
+where
+    G: Scope,
+    G::Timestamp: Lattice + Refines<T>,
+    T: Timestamp + Lattice,
+{
+    Arrangement(Arrangement<G, Row>),
+    CollectionBundle(CollectionBundle<G, Row, T>),
+}
+
+impl<G, T> ArrangementOrBundle<G, T>
+where
+    G: Scope,
+    G::Timestamp: Lattice + Refines<T>,
+    T: Timestamp + Lattice,
+{
+    fn to_bundle(
+        self,
+        key_arity: usize,
+        err_input: Collection<G, DataflowError>,
+    ) -> CollectionBundle<G, Row, T> {
+        match self {
+            ArrangementOrBundle::Arrangement(arrangement) => CollectionBundle::from_columns(
+                0..key_arity,
+                ArrangementFlavor::Local(arrangement, err_input.arrange()),
+            ),
+            ArrangementOrBundle::CollectionBundle(b) => b,
```

Comment:
Pointing out a potential diff in behavior: Previously, all arrangements would be produced with a specific `key_arity`. With this change, that's only the case for reductions maintaining their outputs. Reductions returning a collection would not have the same bundle.